python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -527,3 +479,113 @@ def getExecutionOrder(cls):
+
+        Returns
+        -------
+        badpix : `~galSim.Image`
+            Image representing bad pixels, where zero indicates good pixels and
+            any nonzero value denotes a bad pixel.
+
+        Raises
+        ------
+        MeasurementError
+            If there's an issue during the noise adjustment process.
+        FatalAlgorithmError
+            If BGMEAN is not present in the metadata when using the meta noise
+            source.
+        """
+        # Psf image crossing exposure edge is fine if we're getting the
+        # variance from metadata, but not okay if we're getting the
+        # variance from the variance plane. In both cases, set the EDGE
+        # flag, but only fail hard if using variance plane.
+        overlap = psfImage.getBBox()
+        overlap.clip(exposure.getBBox())
+        if overlap != psfImage.getBBox():
+            self.flagHandler.setValue(record, self.EDGE.number, True)
+            if self.config.noiseSource == "variance":
+                self.flagHandler.setValue(record, self.FAILURE.number, True)
+                raise measBase.MeasurementError(self.EDGE.doc, self.EDGE.number)
+
+        # Match PSF flux to source.
+        psfImage *= record.getPsfInstFlux()
+
+        # Add Gaussian noise to image in 4 steps:
+        # 1. Initialize the noise image and random number generator.
+        noise = afwImage.Image(psfImage.getBBox(), dtype=psfImage.dtype, initialValue=0.0)
+        seed = record.getId() + self.config.seedOffset
+        rand = afwMath.Random("MT19937", seed)
+
+        # 2. Generate Gaussian noise image.
+        afwMath.randomGaussianImage(noise, rand)
+
+        # 3. Determine the noise scaling based on the noise source.
+        if self.config.noiseSource == "meta":
+            # Retrieve the BGMEAN from the exposure metadata.
+            try:
+                bgmean = exposure.getMetadata().getAsDouble("BGMEAN")
+            except NotFoundError as error:
+                raise measBase.FatalAlgorithmError(str(error))
+            # Scale the noise by the square root of the background mean.
+            noise *= np.sqrt(bgmean)
+        elif self.config.noiseSource == "variance":
+            # Get the variance image from the exposure and restrict to the
+            # PSF bounding box.
+            var = afwImage.Image(
+                afwImage.Image(exposure.getMaskedImage().getVariance(), dtype=psfImage.dtype, deep=True),
+                bbox=psfImage.getBBox(),
+                origin=afwImage.PARENT,
+                dtype=psfImage.dtype,
+                deep=False,
+            )
+            # Scale the noise by the square root of the variance.
+            var.sqrt()  # In-place square root.
+            noise *= var
+
+        # 4. Add the scaled noise to the PSF image.
+        psfImage += noise
+
+        # Masking is needed for debiased PSF moments.
+        badpix = afwImage.Mask(psfBBox)
+        # Safer to uncomment the lines below?
```

Comment:
Question: It appears that the `overlap` obtained from these two lines (currently commented out, but implemented in the C++ version) actually yields the same `overlap` as the one calculated earlier at the beginning of the method. To confirm, should I uncomment these lines to precisely mirror the C++ code, or is it logically redundant to repeat the `overlap` calculation?

Reply:
In other words, given that
```
psf = exposure.getPsf()
psfBBox = psf.computeImageBBox(center)
badpix = afwImage.Mask(psfBBox)
if self.config.useSourceCentroidOffset:
    psfImage = psf.computeImage(center)
else:
    psfImage = psf.computeKernelImage(center)
    psfImage.setXY0(psfBBox.getMin())
```
are these the same?
```
overlap = psfImage.getBBox()
```
and
```
overlap = badpix.getBBox()
```

Reply:
I'm only really confident in these being the same for the `True` branch of that `if` statement (though it very well could be true for the `False` branch too). I think I'd recommend uncommenting though.  I think there's zero chance this is a time sink, and using the bbox from the image that's about to be cropped seems safer than using the bbox from a different image, even if they're nominally supposed to have the same bounds.

Reply:
Uncommented those two lines.